src/ui.rs

```diff
@@ -185,8 +270,19 @@ impl Ui {
     }
 
     pub fn flush(&mut self) -> io::Result<()> {
-        match &mut self.output_pair {
-            UiOutputPair::Terminal { stdout, .. } => stdout.flush(),
+        match &mut self.output {
+            UiOutput::Terminal { stdout, .. } => stdout.flush(),
+            UiOutput::Paged { child_stdin, .. } => child_stdin.flush(),
+        }
+    }
+
+    pub fn finalize_writes(&mut self) -> io::Result<()> {
```

Comment:
If this function is intended to close the `ChildStdin`, it needs to take some action to drop it, e.g. by replacing `self.output` with `UiOutput::Terminal` again, as suggested in the adjacent thread.

Reply:
Thanks. I think this is identical to @yuja's suggestion to drop the `ChildStdin`, i.e.

```
    pub fn finalize_writes(&mut self) -> io::Result<()> {
        match &mut self.output {
            UiOutput::Paged { child, child_stdin } => {
                # add this line
                drop(child_stdin);
                child.wait()?;
            }
            _ => (),
        };
        Ok(())
    }
```

Which _should_ fix the issue (since dropping the `ChildStdin` should `close()` the fd), but it doesn't :/

Reply:
It doesn't fix the issue because it doesn't drop the `ChildStdin`, only a reference to it. Because you're matching on `&mut self.output`, any pattern variables are implicitly bound to mutable references to the places they match, rather than being bound to their values. This feature is referred to as "match ergonomics" and I recommend avoiding it entirely due to this type of confusion. Always match on values rather than references, and when you want a reference to a place, request it explicitly with `ref [mut]` qualifiers on the pattern variable.

Reply:
Yeah, I'm not against "match ergonomics", but this one is tricky. Fortunately, clippy can detect noop drop.
```
error: calls to `std::mem::drop` with a reference instead of an owned value. Dropping a reference does nothing
   --> src/ui.rs:282:17
    |
282 |                 drop(child_stdin);
    |                 ^^^^^^^^^^^^^^^^^
    |
    = note: `#[deny(clippy::drop_ref)]` on by default
note: argument has type `&mut std::process::ChildStdin`
   --> src/ui.rs:282:22
    |
282 |                 drop(child_stdin);
    |                      ^^^^^^^^^^^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#drop_ref
```

Reply:
Perfect. Thanks so much for the help everyone!